### PR TITLE
Update 'convert to raw string' to work with interpolated strings.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/65625

The original provider only worked with normal strings (`""` and `@""` strings).  Now we work on interpolated strings as well (`$""` and `$@""` strings).

This required a *lot* of refactoring of the feature as well.  I'll be including a lot of comments to explain everything that is going on.

### CODE REVIEW

File: src/Compilers/Test/Core/MarkedSource/MarkupTestFile.cs [hunk 1]
```diff
@@ -50,7 +50,7 @@ public static class MarkupTestFile
             RegexOptions.Multiline | RegexOptions.IgnorePatternWhitespace);
 
         private static void Parse(
-            string input, out string output, out int? position, out IDictionary<string, ArrayBuilder<TextSpan>> spans)
+            string input, bool treatPositionIndicatorsAsCode, out string output, out int? position, out IDictionary<string, ArrayBuilder<TextSpan>> spans)
```
Comment (CyrusNajmabadi) [inline src/Compilers/Test/Core/MarkedSource/MarkupTestFile.cs:53]: needed to update our internal test harness to allow the test to configure how `$$` is treated.  For raw-string tests, we do not want to treat it as the caret.

File: src/Features/CSharp/Portable/ConvertToRawString/CanConvertParams.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+using Microsoft.CodeAnalysis.CodeActions;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertToRawString;
+
+internal readonly struct CanConvertParams(CodeActionPriority priority, bool canBeSingleLine, bool canBeMultiLineWithoutLeadingWhiteSpaces)
+{
+    public CodeActionPriority Priority { get; } = priority;
+    public bool CanBeSingleLine { get; } = canBeSingleLine;
+    public bool CanBeMultiLineWithoutLeadingWhiteSpaces { get; } = canBeMultiLineWithoutLeadingWhiteSpaces;
+}
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ConvertToRawString/CanConvertParams.cs:14]: this is just a move of this type from an inner type to a top level type.

File: src/Features/CSharp/Portable/ConvertToRawString/ConvertInterpolatedStringToRawStringCodeRefactoringProvider.cs [hunk 3]
```diff
@@ -0,0 +1,685 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ConvertToRawString/ConvertInterpolatedStringToRawStringCodeRefactoringProvider.cs:1]: this is the 'meat' of the change.  it's all the new code to deal with interpolated string conversion.

File: src/Features/CSharp/Portable/ConvertToRawString/ConvertRegularStringToRawStringCodeRefactoringProvider.cs [hunk 4]
```diff
@@ -3,111 +3,44 @@
 // See the LICENSE file in the project root for more information.
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ConvertToRawString/ConvertRegularStringToRawStringCodeRefactoringProvider.cs:3]: i kep the name of this file the same to keep the diff clear.    the CodeREfactoringProvider code was in here, as well as the code to deal with normal (non-interpolated) strings.

Now, this file just contains the code to deal with the strings.  Everything related to the CodeRefactoringProvider moved into a separate file.  The CodeRefactoringProvider now calls into the helpers for normal strings and interpolated strings to do the actual processing.

File: src/Features/CSharp/Portable/ConvertToRawString/ConvertRegularStringToRawStringCodeRefactoringProvider.cs [hunk 5]
```diff
@@ -3,111 +3,44 @@
...
-    internal partial class ConvertRegularStringToRawStringCodeRefactoringProvider : SyntaxEditorBasedCodeRefactoringProvider
+    internal partial class ConvertRegularStringToRawStringProvider
+        : AbstractConvertStringProvider<LiteralExpressionSyntax>
     {
-        private enum ConvertToRawKind
-        {
-            SingleLine,
-            MultiLineIndented,
-            MultiLineWithoutLeadingWhitespace,
-        }
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ConvertToRawString/ConvertRegularStringToRawStringCodeRefactoringProvider.cs:40]: moved to top level type.

File: src/Features/CSharp/Portable/ConvertToRawString/ConvertRegularStringToRawStringCodeRefactoringProvider.cs [hunk 6]
```diff
@@ -3,111 +3,44 @@
...
+        public static readonly IConvertStringProvider Instance = new ConvertRegularStringToRawStringProvider();
 
-        private static readonly BidirectionalMap<ConvertToRawKind, string> s_kindToEquivalenceKeyMap =
-            new(new[]
-            {
-                KeyValuePairUtil.Create(ConvertToRawKind.SingleLine, nameof(ConvertToRawKind.SingleLine)),
-                KeyValuePairUtil.Create(ConvertToRawKind.MultiLineIndented, nameof(ConvertToRawKind.MultiLineIndented)),
-                KeyValuePairUtil.Create(ConvertToRawKind.MultiLineWithoutLeadingWhitespace, nameof(ConvertToRawKind.MultiLineWithoutLeadingWhitespace)),
-            });
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ConvertToRawString/ConvertRegularStringToRawStringCodeRefactoringProvider.cs:48]: moved to coderefactoringprovider.

File: src/Features/CSharp/Portable/ConvertToRawString/ConvertRegularStringToRawStringCodeRefactoringProvider.cs [hunk 7]
```diff
@@ -3,111 +3,44 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ConvertToRawString/ConvertRegularStringToRawStringCodeRefactoringProvider.cs:5]: review with whitespace off.

File: src/Features/CSharp/Portable/ConvertToRawString/ConvertRegularStringToRawStringCodeRefactoringProvider.cs [hunk 8]
```diff
@@ -3,111 +3,44 @@
...
-        public ConvertRegularStringToRawStringCodeRefactoringProvider()
+        private ConvertRegularStringToRawStringProvider()
         {
         }
 
-        protected override ImmutableArray<FixAllScope> SupportedFixAllScopes => AllFixAllScopes;
+        protected override bool CheckSyntax(LiteralExpressionSyntax stringExpression)
+            => stringExpression.Kind() is SyntaxKind.StringLiteralExpression;
 
-        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ConvertToRawString/ConvertRegularStringToRawStringCodeRefactoringProvider.cs:58]: this code moved to the new file, which is the actual coderefactoringprovider.

File: src/Features/CSharp/Portable/ConvertToRawString/ConvertRegularStringToRawStringCodeRefactoringProvider.cs [hunk 9]
```diff
@@ -167,100 +96,41 @@ private static bool CanConvertStringLiteral(SyntaxToken token, out CanConvertPar
...
-                    index--;
-
-                return index >= 0 && IsCSharpNewLine(characters[index]);
-            }
-        }
+            // a single line raw string cannot contain a newline.
+            if (characters.Any(static ch => IsCSharpNewLine(ch)))
+                return false;
 
-        private static async Task<Document> UpdateDocumentAsync(
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ConvertToRawString/ConvertRegularStringToRawStringCodeRefactoringProvider.cs:192]: UpdateDocumentAsync moved to coderefactoringprovider

File: src/Features/CSharp/Portable/ConvertToRawString/ConvertRegularStringToRawStringCodeRefactoringProvider.cs [hunk 10]
```diff
@@ -167,100 +96,41 @@ private static bool CanConvertStringLiteral(SyntaxToken token, out CanConvertPar
...
-            Contract.ThrowIfFalse(span.IntersectsWith(token.Span));
-            Contract.ThrowIfFalse(token.Kind() == SyntaxKind.StringLiteralToken);
-
-            var parsedDocument = await ParsedDocument.CreateAsync(document, cancellationToken).ConfigureAwait(false);
-            var replacement = GetReplacementToken(parsedDocument, token, kind, options, cancellationToken);
-            return document.WithSyntaxRoot(root.ReplaceToken(token, replacement));
+            return true;
         }
 
-        protected override async Task FixAllAsync(
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ConvertToRawString/ConvertRegularStringToRawStringCodeRefactoringProvider.cs:206]: UpdateDocumentAsync moved to coderefactoringprovider

File: src/Features/CSharp/Portable/ConvertToRawString/ConvertStringToRawStringCodeRefactoringProvider.cs [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertToRawString;
+
+[ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = PredefinedCodeRefactoringProviderNames.ConvertToRawString), Shared]
+internal partial class ConvertStringToRawStringCodeRefactoringProvider : SyntaxEditorBasedCodeRefactoringProvider
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ConvertToRawString/ConvertStringToRawStringCodeRefactoringProvider.cs:25]: this is now the actual refactoring provider.  it is a move of all teh code from the other file.

File: src/Features/CSharp/Portable/ConvertToRawString/ConvertStringToRawStringCodeRefactoringProvider.cs [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+        new(new[]
+        {
+            KeyValuePairUtil.Create(ConvertToRawKind.SingleLine, nameof(ConvertToRawKind.SingleLine)),
+            KeyValuePairUtil.Create(ConvertToRawKind.MultiLineIndented, nameof(ConvertToRawKind.MultiLineIndented)),
+            KeyValuePairUtil.Create(ConvertToRawKind.MultiLineWithoutLeadingWhitespace, nameof(ConvertToRawKind.MultiLineWithoutLeadingWhitespace)),
+        });
+
+    private static readonly ImmutableArray<IConvertStringProvider> s_convertStringProviders = ImmutableArray.Create<IConvertStringProvider>(
+        ConvertRegularStringToRawStringProvider.Instance,
+        ConvertInterpolatedStringToRawStringProvider.Instance);
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ConvertToRawString/ConvertStringToRawStringCodeRefactoringProvider.cs:37]: here are the two providers for handling the two types of strings.

**Comment by CyrusNajmabadi** [comment]
@akhera99 @genlu ptal.

**Comment by DoctorKrolic** [comment]
Closes https://github.com/dotnet/roslyn/issues/65625

**Comment by CyrusNajmabadi** [comment]
@akhera99 @genlu ptal.

**Comment by CyrusNajmabadi** [comment]
@akhera99 @genlu ptal.

**Review by akhera99 — APPROVED** [review]

**Pull request merged**